reconcile/openshift_namespace_labels.py

```diff
@@ -0,0 +1,395 @@
+        """Add an error to the given cluster / namespace"""
+        self.errors(cluster, namespace).append(err)
+
+    def has_any_error(self) -> bool:
+        """Checks if any cluster / namespace has any error registered"""
+        return any(e[2] for e in self.iter_errors())
+
+    def iter_errors(self) -> Generator[Tuple[str, str, List[str]], None, None]:
+        """yields (cluster, namespace, errors) items"""
+        for cluster, namespaces in self._errors.items():
+            for namespace, errors in namespaces.items():
+                if errors:
+                    yield cluster, namespace, errors
+
+    def _ns(self, cluster: str, namespace: str) -> Types:
+        return self._inv.setdefault(cluster, {}).setdefault(namespace, {})
+
+    def get(self, cluster: str, namespace: str, type: str,
+            default: Optional[LabelsOrKeys] = None) -> Optional[LabelsOrKeys]:
+        """Get the labels or keys for the given cluster / namespace / type"""
+        return self._inv.get(cluster, {}).get(namespace, {}).get(type, default)
+
+    def setdefault(self, cluster: str, namespace: str, type: str,
+                   default: LabelsOrKeys) -> LabelsOrKeys:
+        """Get the labels or keys for the given cluster / namespace / type,
+        setting it to default if it does not exists"""
+        with self._lock:
+            return self._ns(cluster, namespace).setdefault(type, default)
+
+    def set(self, cluster: str, namespace: str,
+            type: str, labels: Labels) -> Labels:
+        """Sets the given cluster / namespace / type to 'labels'"""
+        with self._lock:
+            self._ns(cluster, namespace)[type] = labels
+            return labels
+
+    def delete(self, cluster: str, namespace: str):
+        """Delete the given cluster / namespace from the inventory"""
+        with self._lock:
+            self._inv.get(cluster, {}).pop(namespace, None)
+
+    def __iter__(self) -> Generator[Tuple[str, str, Types], None, None]:
+        """Makes the inventory iterable by yielding (cluster, namespace, types)
+        items. Types here is a dict of {type: labelsOrKeys} """
+        for cluster, namespaces in self._inv.items():
+            for namespace, types in namespaces.items():
+                yield cluster, namespace, types
+
+    def update_managed_keys(self, cluster: str, namespace: str, key: str):
+        """
+        Add or remove a key from the managed key list.
+        This actually handles a copy of the managed keys dict and updates it.
+        If the key was managed, it will get removed.
+        If it was not, it will get added
+        """
+        managed = self.get(cluster, namespace, MANAGED, [])
+        if managed is None:
+            managed = []
+        upd_managed = self.setdefault(cluster, namespace, UPDATED_MANAGED,
+                                      managed.copy())
+
+        assert isinstance(upd_managed, list)  # we never get a Dict here
```

Comment:
All of the comments below are just ideas how we might be able to avoid this case in the future.  I am not suggesting that they are necessary for this PR at all.  The occasional `assert` where we know that something truly can't happen is reasonable, but I'd be more concerned if this becomes a regular thing.

For what it's worth, I found: https://pypi.org/project/typing-extensions/

That allows us to import `TypedDict` ([docs](https://docs.python.org/3/library/typing.html#typing.TypedDict)) and others more recent features from the `typing` module before we upgrade to 3.9.  I think that I'll go ahead and add that to my Mypy PR.  A `TypedDict` would provide the link between the `UPDATE_MANAGED` key and the `list` type, compared the other keys that are returning `dict`.

It is worth mentioning that `TypedDict`s are fairly strict by design.  You'd need to use `total=False` if all of the keys aren't guaranteed to be present and you can't use a variable typed as a `str` during a lookup without an error, so you'd need to do something like:

```
StateKeys = Literal['desired', 'managed', 'current', 'changed',
                    'update_managed']
```

Thanks for using this as an opportunity to add types to this new integration!

Reply:
Just FYI, I went down the rabbit hole and even with a `TypedDict`, I get the same error message if I remove your `assert`:

```
reconcile/openshift_namespace_labels.py:135: error: Item "Dict[str, Optional[str]]" of "Union[Dict[str, Optional[str]], List[str]]" has no attribute "remove"
reconcile/openshift_namespace_labels.py:137: error: Item "Dict[str, Optional[str]]" of "Union[Dict[str, Optional[str]], List[str]]" has no attribute "append"
```

The Types dict would look like:

```
class Types(TypedDict, total=False):
    desired: Labels
    managed: LabelKeys
    current: Labels
    changed: Labels
    updated_managed: LabelKeys
```


But if you change the code to:

```
        managed = self._inv.get(cluster, {}).get(namespace, {}).get('managed')
        if managed is None:
            managed = []
        upd_managed = self._ns(cluster, namespace).setdefault(
            "updated_managed", managed.copy())
```

Then, Mypy is happy (with this issue, but there were a few others).  So, basically if the method can return A or B, and the `TypedDict` indicates that given a certain key A is returned, Mypy will still assume A or B because that's what your method signature indicates.  You'd need to directly access the `TypedDict` from what I'm seeing (kind of makes sense).

I'll dig into this again when I have some free time because I think these sorts of issues are going to be where we feel the most pain.  We may just find `dataclass` is the way to go where the level of nesting is not large and the attributes are well-known.